File: website/web/src/api/statistics.ts
/**
 * @Author     : Amu
 * @Date       : 2025/02/12 16:58:24
 * @Description:
 */

import type { StatisticsQueryReply, StatisticsUpdateParams, StatisticsUpdateReply } from '~/interface/statistics'

import { useHttp } from '~/composables/useHttp'
import { API } from '~/config/api'

export async function statisticQuery() {
    const reply = await useHttp.get<StatisticsQueryReply>(API.statistics_query, {}, { silent: true })
    const statistic = reply?.data
    if (
        !statistic
        || !Number.isInteger(statistic.id)
        || statistic.id <= 0
        || !Number.isInteger(statistic.times)
        || statistic.times < 0
    ) {
        throw new Error('统计响应格式错误')
    }
    return statistic
}

export async function statisticUpdate(params: StatisticsUpdateParams) {
    try {
        return await useHttp.post<StatisticsUpdateReply>(API.statistics_update, params)
    }
    catch (e) {
        console.error('[API Error] statisticUpdate:', e)
        throw new Error('数据更新失败')
    }
}

File: website/server/service/statistics/service/statistics.go
// Package service
// Date: 2025/02/12 15:25:53
// Author: Amu
// Description:
package service

import (
	"context"
	"github.com/google/wire"
	"server/service/schema"
	"server/service/statistics/repository"
)

var StatisticsServiceSet = wire.NewSet(NewStatisticsService, wire.Bind(new(IStatisticsService), new(*StatisticsService)))

var _ IStatisticsService = (*StatisticsService)(nil)

type IStatisticsService interface {
	StatisticsQuery(context.Context) (schema.StatisticsQueryReply, error)
	StatisticsUpdate(context.Context, schema.StatisticsUpdateArgs) (schema.StatisticsUpdateReply, error)
	InstallationReport(context.Context, schema.InstallationReportArgs) (schema.InstallationReportReply, error)
}

type StatisticsService struct {
	StatisticsRepository repository.IStatisticsRepository
}

func NewStatisticsService(repo repository.IStatisticsRepository) *StatisticsService {
	return &StatisticsService{StatisticsRepository: repo}
}

func (s *StatisticsService) StatisticsQuery(ctx context.Context) (schema.StatisticsQueryReply, error) {
	result := schema.StatisticsQueryReply{}
	reply, err := s.StatisticsRepository.StatisticsQuery(ctx)
	if err != nil {
		return result, err
	}
	result.Data.ID = reply.ID
	result.Data.Times = reply.Times
	return result, nil
}

func (s *StatisticsService) StatisticsUpdate(ctx context.Context, args schema.StatisticsUpdateArgs) (schema.StatisticsUpdateReply, error) {
	result := schema.StatisticsUpdateReply{}
	return result, s.StatisticsRepository.StatisticsUpdate(ctx, args)
}

func (s *StatisticsService) InstallationReport(ctx context.Context, args schema.InstallationReportArgs) (schema.InstallationReportReply, error) {
	result := schema.InstallationReportReply{}
	return result, s.StatisticsRepository.InstallationReport(ctx, args)
}
